File: jugglerhelper/src/main/java/net/maxsmr/jugglerhelper/fragments/splash/BaseSplashJugglerFragment.java
package net.maxsmr.jugglerhelper.fragments.splash;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import org.jetbrains.annotations.Nullable;
import android.view.View;

import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment;


public abstract class BaseSplashJugglerFragment extends BaseJugglerFragment {

    private static final String ARG_EXPIRED_TIME = BaseSplashJugglerFragment.class.getName() + ".ARG_EXPIRED_TIME";

    private final Handler navigateHandler = new Handler(Looper.getMainLooper());
    private final Runnable navigateRunnable = new Runnable() {
        @Override
        public void run() {
            if (isAdded()) {
                onSplashTimeout();
            }
        }
    };

    private long expiredTime = 0;
    private long startTime = 0;

    protected abstract long getSplashTimeout();

    protected abstract boolean allowResetOnRootClick();

    protected abstract void onSplashTimeout();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        expiredTime = savedInstanceState != null ? savedInstanceState.getLong(ARG_EXPIRED_TIME) : expiredTime;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (allowResetOnRootClick()) {

            View rootView = getView();

            if (rootView == null) {
                throw new RuntimeException("root view was not inflated");
            }

            rootView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    navigateHandler.removeCallbacks(navigateRunnable);
                    onSplashTimeout();
                }
            });
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        long timeout = getSplashTimeout();

        if (timeout < 0) {
            throw new IllegalArgumentException("incorrect splash timeout");
        }

        startTime = System.currentTimeMillis() - expiredTime;
        navigateHandler.postDelayed(navigateRunnable, timeout - expiredTime);
    }

    @Override
    public void onStop() {
        super.onStop();
        navigateHandler.removeCallbacks(navigateRunnable);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong(ARG_EXPIRED_TIME, expiredTime = (startTime > 0 ? System.currentTimeMillis() - startTime : 0));
    }

}

File: jugglerhelper/src/main/kotlin/net/maxsmr/jugglerhelper/fragments/splash/BaseSplashJugglerFragment.kt
package net.maxsmr.jugglerhelper.fragments.splash

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.annotation.IdRes
import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment

private val ARG_EXPIRED_TIME = BaseSplashJugglerFragment::class.java.name + ".ARG_EXPIRED_TIME"

abstract class BaseSplashJugglerFragment : BaseJugglerFragment() {

    protected abstract val splashTimeout: Long

    @IdRes
    protected open val layoutClickResId = 0

    private val navigateRunnable = Runnable {
        if (isAdded) {
            onSplashTimeout()
        }
    }
    private var expiredTime: Long = 0

    private var startTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        expiredTime = savedInstanceState?.getLong(ARG_EXPIRED_TIME) ?: expiredTime
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        with(layoutClickResId) {
            if (this != 0) {
                view.findViewById<View>(this)?.setOnClickListener {
                    mainHandler.removeCallbacks(navigateRunnable)
                    onSplashTimeout()
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        val timeout = splashTimeout
        require(timeout >= 0) { "incorrect splash timeout" }
        startTime = System.currentTimeMillis() - expiredTime
        mainHandler.postDelayed(navigateRunnable, timeout - expiredTime)
    }

    override fun onStop() {
        super.onStop()
        mainHandler.removeCallbacks(navigateRunnable)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(ARG_EXPIRED_TIME, (if (startTime > 0) System.currentTimeMillis() - startTime else 0).also { expiredTime = it })
    }

    protected abstract fun onSplashTimeout()
}